### ISSUE TITLE
Breaking up the Transaction class

### ISSUE THREAD

**Issue by s1na** [ethereumjs/ethereumjs-tx#151]
Created this issue to continue the discussion started in #142 and to garner feedback whether you think this'd make sense. Here's a rough sketch of one (non-breaking) approach that comes to my mind:

Make [TxData](https://github.com/ethereumjs/ethereumjs-tx/blob/master/src/types.ts#L24) a class, and use it to store the actual data for the transaction, plus methods for serialization/deserialization. In `Transaction`'s constructor instantiate `TxData` and store it for later use (possibly add a `static fromTxData` method). Logic methods stay on `Transaction` and use `TxData` as source of data but do not mutate it.

If this is too vague I can try to write up (or a diagram) of something more concrete. Please feel free to propose other ideas.

**Comment by alcuadrado** [comment]
I see the issues in #140 and #150 related to this one, and that's why I proposed moving those conversations here.

IMO the problems we are seeing come from a few design desitions:

1) The internal representation of the transaction values is completely coupled to RLP. This leads to confusing things, like leading 0s being removed unexpectedly.

2) Unsigned transactions have bogus signatures. This is confusing both as a developer of the lib and as a user of it, for example `tx.hash(false)` is weird.

3) Many operations mutate the internal state and these side effects are used internally in unexpected ways. A clear example of this is `Transaction.prototype.getSenderPublicKey`'s implementation.



> Created this issue to continue the discussion started in #142 and to garner feedback whether you think this'd make sense. Here's a rough sketch of one (non-breaking) approach that comes to my mind:
> 
> Make [TxData](https://github.com/ethereumjs/ethereumjs-tx/blob/master/src/types.ts#L24) a class, and use it to store the actual data for the transaction, plus methods for serialization/deserialization. In `Transaction`'s constructor instantiate `TxData` and store it for later use (possibly add a `static fromTxData` method). Logic methods stay on `Transaction` and use `TxData` as source of data but do not mutate it.
> 
> If this is too vague I can try to write up (or a diagram) of something more concrete. Please feel free to propose other ideas.

What I don't understand is how to make this in a non-breaking way.

**Comment by s1na** [comment]
> What I don't understand is how to make this in a non-breaking way.

Hm, you're right. It could be possible to have it non-breaking by having getters for values on `Transaction`. But that won't be an improvement to the current design...

@alcuadrado What do you propose? I think some breaking change is inevitable.

**Comment by alcuadrado** [comment]
A good way to move forward may be to do a clean redesign of the library, and then evaluate if it makes sense to replace it entirely or only we want part of the changes. There's the possibility of some work not ending up in the library, but we get to experiment with solving all those things.

I can do this if we agree that it's something that we want to do.

@danjm, you mentioned in #142 that you had some ideas about redesigning this lib. Mind sharing them? Thanks!

**Comment by rajeshsubhankar** [comment]
I have 2 use cases where I struggled to understand how the library works.

### Use case 1 (no EIP-155 support):
I want to get the `rlp` encoded serialized transaction data without `v,r,s` which I will hash and sign in some different platform.
__expectation__: `tx.serialize(false).toString('hex')` 
__currently__: `RLP.encode(tx.raw.slice(0, 6)).toString('hex')`

Once I receive the signature, I should be able to update my transaction easily.
__expectation__: `tx.updateSignature(sig)`
__currently__: 
```js
tx.r: sig.signature.slice(0, 32).toString('hex);
tx.s: sig.signature.slice(32, 64).toString('hex);
tx.v: sig.recovery + 27;
```
Where `const sig = secp256k1.sign(txData, privKey)`
Else, we can restrict `sig` to be strictly `DER` encoded as specified in [BIP66](https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki)

### Use case 2 (EIP-155 support)
Create a raw transaction with `v=3`,`r=0` and `s=0`, get the `rlp` encoded serialized transaction data by `tx.serialize().toString('hex')` then hash and sign it in some different platform.

Now, when I update the `tx` object manually by adding the `v,r,s` from the received `sig` with appropriate `v` as >=37, I always get an invalid signature error.

Unfortunately, `verifySignature()` method always does `const msgHash =  this.hash(false)` ([transaction.ts#L260](https://github.com/ethereumjs/ethereumjs-tx/blob/3648da44355e7f6f6c10bdbbd3fc0c8fd1e17abc/src/transaction.ts#L260)), which depends on `chainId`(which I didn't specify while creating the raw tx) to construct the correct hash.

So, it's confusing whether to specify `chainId` in the raw transaction in the first place or to modify `v` directly with the correct chainId value.

I think the library is assuming that for every external sign use case, the user will specify the `chainId` and take the `tx.hash(false)` rather than modifying the `v` and taking the `tx.serialize()` .

**Comment by holgerd77** [comment]
Hi @rajeshsubhankar, your comment had absolutely nothing to do with the issue discussed here, please make sure that you do comments or open issues where it fits.

Have deleted your post and re-posted as a new issue, see above.

**Comment by alcuadrado** [comment]
I agree with moving that to a different issue, but we should note that what @rajeshsubhankar describes as "Use case 2" is important for this discussion. 

Basically, he is having problems with setting and verifying a signature after constructing the TX object. The problem is that the TX object only computes the chainId on construction, not on the `v` setter.

**Comment by alcuadrado** [comment]
Hey @s1na, I created PR #154 to explore one possibility of what we can do with this and other libraries.  

The idea of that PR is to keep everything as explicit as possible. Transaction APIs now work with buffers, with the only exception of static factory methods added for the users' convenience.

In this new version, transactions aren't immutable either, but only `sign` and the setters mutate them.

It has about 60 lines more of boilerplate, but I think it's a fair price for making it much easier to understand and work with.
